### PR TITLE
Feat/filtration3

### PR DESCRIPTION
**Pull request by jedrzejSmok**

### CODE REVIEW

File: src/main/java/pl/poznan/put/jsontools/logic/JsonTransformRetainAttributes.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+     * @return  JSON containing only attributes specified by  attributes
+     */
+    @Override
+    public String execute() {
+        String json = super.execute();
+        _logger.debug("Executing saving attributes");
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode jsonnode = mapper.readTree(json);
+            List<String> jsonattributes = new ArrayList<>();
```
Comment (pawelstrzelczyk) [inline src/main/java/pl/poznan/put/jsontools/logic/JsonTransformRetainAttributes.java:50]:
```suggestion
            List<String> jsonAttributes = new ArrayList<>();
```

File: src/main/java/pl/poznan/put/jsontools/logic/JsonTransformRetainAttributes.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+     * It retains given attributes.
+     * @return  JSON containing only attributes specified by  attributes
+     */
+    @Override
+    public String execute() {
+        String json = super.execute();
+        _logger.debug("Executing saving attributes");
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode jsonnode = mapper.readTree(json);
```
Comment (pawelstrzelczyk) [inline src/main/java/pl/poznan/put/jsontools/logic/JsonTransformRetainAttributes.java:49]:
```suggestion
            JsonNode jsonNode = mapper.readTree(json);
```

**Review by pawelstrzelczyk — CHANGES_REQUESTED** [review]
works fine ;-)

**Review by pawelstrzelczyk — APPROVED** [review]
thanks

**Review by pawelstrzelczyk — APPROVED** [review]

File: src/main/java/pl/poznan/put/jsontools/logic/JsonTransformRetainAttributes.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        String json = super.execute();
+        _logger.debug("Executing saving attributes");
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode jsonNode = mapper.readTree(json);
+            List<String> jsonAttributes = new ArrayList<>();
+            Iterator<String> NameIterator = jsonNode.fieldNames();
+            while(NameIterator.hasNext()) {
+                jsonAttributes.add(NameIterator.next());
+            }
```
Comment (codeblessing) [inline src/main/java/pl/poznan/put/jsontools/logic/JsonTransformRetainAttributes.java:54]: Instead use:
```suggestion
            List<String> jsonAttributes = new ArrayList<>();
            jsonNode.fieldNames().forEachRemaining(jsonAttributes::add);
```

File: src/main/java/pl/poznan/put/jsontools/logic/JsonTransformRetainAttributes.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            for(String attribute: jsonAttributes){
+                if(!this._attributes.contains(attribute)){
+                    ((ObjectNode)jsonNode).remove(attribute);
+                }
+            }
+
+            return mapper.convertValue(jsonNode, JsonNode.class).toString();
```
Comment (codeblessing) [inline src/main/java/pl/poznan/put/jsontools/logic/JsonTransformRetainAttributes.java:61]: That's unnecessary. `jsonNode` is already JsonNode type.
`return jsonNode.toString()` should be enough.

**Review by codeblessing — CHANGES_REQUESTED** [review]

**Review by codeblessing — APPROVED** [review]

**Review by Mmajonez — APPROVED** [review]

**Pull request merged**